### DISCUSSION TITLE
Dynamic Generics

### DISCUSSION THREAD

**Discussion by MiddleTommy** [dotnet/csharplang#7916] [Language Ideas]
I often wish generics in c# would work like this.
```
var mytype = MyObject.GetType();
var list = new List<mytype>();
```
Often you can't use generics because you don't know the type at compile time. Who would like to have all the reflection code to make this work, generated for you at compile time?

**Comment by tacosontitan** [comment]
If you don't like writing the reflection, then there are some options:

- Refactor so the type is known at compile time.
- Create an extension method.
- Create a source generator.

I can't say there has ever been a situation that I had control over in which I couldn't know the type at compile time. Consequently, since this is something you encounter often, can you provide a use case in which the type couldn't possibly be known at compile time?

**Reply by MiddleTommy** [reply]
I am not saying that there are not work arounds. This is where 18 years ago as a new programmer I would always get stuck. I would not have the reflection expertise to do this

`var typecache = typeof(TypeCache<>).MakeGenericType(doctype);
var cache = Activator.CreateInstance(typecache);`

Then with parameters on the constructor it gets more complicated. 
How about using dynamic to call a generic method?

```
dynamic obj = doc.ToObject(t);
DocUpdated(obj);

 public void DocUpdated<T>(T doc)
     where T : TDocument
 {
////code
 }
```
These are all things that are just annoying as a language. Yes there are ways but they are ugly. We also could manually create expression trees (and I do sometimes) but LINQ is easier, less error prone and pretty to see. 

I deal with hundreds of document data types and run into this a few times a year. But it took me years of experience to develop the skills to get past those issues. Now that C# has everything needed, I am asking for features I want that aren't needed.

**Reply by tacosontitan** [reply]
> This is where 18 years ago as a new programmer I would always get stuck.

If you were forced to use reflection as a junior, without guidance, then IMHO your mentors and peers failed you.

> Yes there are ways but they are ugly.

Can you elaborate on ways to overcome reflection that you consider *ugly*?

---
As a whole, your response still doesn't give us a use case that couldn't be refactored to use generics. Both examples you gave immediately had me asking questions:

- Why isn't the type of `doctype` not known upfront?
- Why are you using `dynamic`?

Without a simple but complete and sound example demonstrating a scenario where it is impossible to know the type upfront, it's pretty difficult to understand why you would want to introduce implicit behavior for something that *should* be explicit as often as possible.

**Reply by colejohnson66** [reply]
I agree. This sounds like an X/Y problem.

**Reply by MiddleTommy** [reply]
I am 100% self-taught. I had no mentors, just documentation and blogs. I wish I had YouTube tutorials like there are now.

**Reply by MiddleTommy** [reply]
The type is often not known upfront when you are writing code that others will use. The type may not even be in the same dll your code is in.
Say you want to do just one API call to get many different objects/documents from the server. Your database is a document database and isnt separated by tables. In a document database all your different document types are in the same "table" 
It is much more efficient to request all the documents with a groupId = "xyz" than to request each document type with a groupid (one database call vs multiple).
This will give you a response with different Json object types that you then have to serialize into their proper types. 
Then you return an IEnumerable<object> with multiple object types. If I wanted to write code to place these documents into a separate repository for each type it would be cumbersome reflection. 
The point being there is situations where there is no way to know the class types upfront. Especially if you want to release this code on GitHub for strangers to use.

**Reply by tacosontitan** [reply]
> The type is often not known upfront when you are writing code that others will use.

This is a self-imposed limitation because that's the beauty in generics; you can write code that applies to any type since you don't care about the implementation details the consumer is concerned with. Take a simple serialization provider; I don't care what type the consumer is using because that doesn't impact my efforts to serialize or deserialize it:

```cs
public interface ISerializationProvider {
    string? Serialize<T>(T input);
    T? Deserialize(string input);
}
```

> Say you want to do just one API call to get many different objects/documents from the server.

I'll tell you what I tell everyone else:

![PaulRuddDoLessGIF](https://github.com/dotnet/csharplang/assets/65432314/d29d678f-9f1f-429e-9cab-cc2402ff94a4)

> Your database is a document database and isn't separated by tables.

Okay? You can still organize your data with collections (e.g. a user document collection) and query a concrete type for the documents in that collection. It wouldn't be a good idea (nor very efficient) to query users and vehicles at the same time, so why would you? In my honest opinion, that's something document databases have power over relational ones with (you can query structured data without mapping from flat result sets).

> It is much more efficient to request all the documents with a groupId = "xyz" than to request each document type with a groupid (one database call vs multiple).

That all depends on context. I believe that too many people fall into the preoptimization trap of over-reducing I/O calls before they even determine the actual impact. There are plenty of ways to optimize I/O calls without having to combine concerns and force yourself to write harder-to-maintain code for more often than not, negligible benefits.

> Then you return an `IEnumerable` with multiple object types.

*...never have I ever.*

> The point being there is situations where there is no way to know the class types upfront. Especially if you want to release this code on GitHub for strangers to use.

Generics were designed for exactly that type of versatility, and if you need to know *exactly* what your consumers are sending you, then you're using generics incorrectly in my honest opinion.

**Reply by MiddleTommy** [reply]
@tacosontitan 
All I can say is your arguments are good if you get paid by the hour to write code.

**Reply by tacosontitan** [reply]
> I am 100% self-taught. I had no mentors, just documentation and blogs.

 I'm 100% self-taught myself, but that doesn't mean that there weren't others around me to tell me "that's stupid, try something else" or "look into this". If you've worked in a professional capacity for any length of time and had no mentors, then your peers failed you, or you failed yourself by not letting them mentor you.

**Reply by tacosontitan** [reply]
> All I can say is your arguments are good if you get paid by the hour to write code.

I'm salary, but that doesn't mean I have to write slam code all the time. I slam in emergencies, but those are exceptions, not the norm. I'm not afraid to tell my superiors when something isn't possible to do in a maintainable way in the timeline they want, and they advise me from there. Deadlines are not a reason to write bad code.

> Only you can prevent dumpster fires.

**Reply by MiddleTommy** [reply]
Look @tacosontitan. I am a solo self-taught developer and have never talked to anyone in person ever in my life and only a couple times online about coding. It has always just been me and no-one around me cares. So please tone down the best practices hocus pocus. I just want to get things working as quick as possible and refactor later if needed. You think my idea is bad. thanks, move on.

**Reply by tacosontitan** [reply]
I'm not on "best practices hocus pocus" because I think best practices can constrain a developer too much; they're guidelines, not laws. I'm simply trying to get you to give us a concrete example (with code) that would support this so that we can better understand your idea. If I've offended you, then I apologize, but I'm not attacking you, I'm simply being blunt.

**Comment by TahirAhmadov** [comment]
I can understand what is requested, and I can even imagine situations where this would be needed. But:
1. These situations are very uncommon from the perspective of the overall ecosystem;
2. The proposed syntax is very confusing - language needs to interpret whether the provided identifier is a type or a variable holding a `Type`, and if you consider that it could be a property, it becomes difficult to read - is `ElementType` in `List<ElementType>` a type or a property named `ElementType`?
3. In addition to 2 above - do we also support expression? Then it can become really unwieldy, like `List<someVar.Prop.Foo()>`
4. The readily available API call is not that complicated to warrant a language improvement.

**Reply by tacosontitan** [reply]
Without spending *too* much time thinking about syntax possibilities, *if* we supported such a feature (and I'm not in support right now), then I'd prefer the syntax to look something like this:
```cs
var collection = new from(List<>, elementType);

// short for:
var collection = new from(typeof(List<>), elementType);
```
I initially thought of using `typeof` directly, but that would require giving multiple expectations of the result coming back from `typeof` and that just didn't sit well with me. The next best option for me mentally was `from` because it already exists as a part of query syntax so we wouldn't need a keyword; however, I feel like there's some grey area on if this causes a change in expectation based on context, which again, would not sit well with me.

**Reply by MiddleTommy** [reply]
Im not a fan of "new from" as this is different than normal c# code and doesnt work for generic methods.
I would prefer the @ sign to distinguish the runtime type in the generic arguments instead of compile time types. 
```
Type objtype = MyObject.GetType()
var list = new List<@objtype>();
//or
var result = MethodCall<@objtype>(MyObject);
```

**Reply by HaloFour** [reply]
C# is not a dynamic language.  To make something like that work would require emitting extremely different code, none of which could be statically checked at compile time.  Such things are better suited to dynamic languages designed for that purpose.

What you could do in C# is use reflection to create the types, and then use `dynamic` to write the code that interacts with the created instances.  It's not quite as flexible as a dynamic language would be, but it provides that runtime dispatch.

**Reply by MiddleTommy** [reply]
@HaloFour C# is a static language with dynamic features. We should be improving those dynamic features so when needed it is less of a pain to use c#. Yes, dynamic C# is needed.

**Reply by HaloFour** [reply]
> We should be improving those dynamic features so when needed it is less of a pain to use c#.  Yes, dynamic C# is needed.

Or, you could use one of many available dynamic languages that interops with the CLR.  C# isn't trying to be every language, and it's not a goal to become a dynamic language.

**Reply by MiddleTommy** [reply]
@HaloFour Why would I want to include a dependency on another language and runtime just to process a few lines of code I could write in reflection? Yes, this is not a needed feature. It is a want, to solve a few situations. However, I feel it would be used more often if available. Right now we avoid these situations because they are painful. So, we write way more code to cover Type safety.

At my real job we process steel. We used to cut all plates out of standard width bar sizes (1 inch to 12 inches wide in 1/2 inch increments) because cutting plate manually by hand was time consuming. Plates would be sticking out or inset a little when welded into other steel shapes. We didn't like curves or shapes in plate and forced everything to be only rectangles. Then we got a CNC plasma plate cutter. We still use bar size widths but also cut plates to any size or shape with ease and speed. We can even cut bar sizes out of bigger plates. However, we had to spend $150K to get that machine.

I feel features like this one would be like getting that plasma plate cutting machine for C#. We use the types as often as we can but when that odd shape or situation comes around it would be no problem. 

Generics are dynamic in their nature. Its time we make dynamic generics.

**Reply by HaloFour** [reply]
> Why would I want to include a dependency on another language and runtime

Probably the same reason a toolbox generally contains more than one tool.

> However, I feel it would be used more often if available.

Brittle and slow should not be encouraged.  C# already offers runtime binding to simplify working with dynamic code via the DLR and `dynamic`.

> Generics are dynamic in their nature.

Generics are definitely not dynamic in their nature.

**Reply by MiddleTommy** [reply]
@HaloFour I would agree consuming generic objects is not dynamic. However, when authoring generic classes you often need to think in dynamic ways.

**Reply by tacosontitan** [reply]
> Im not a fan of "new from" as this is different than normal c# code and doesnt work for generic methods.

Your example is *not* leveraging generics...

```cs
public interface ISample { int Value { get; set; } }

...
public bool Foo<TIn>(TIn input)
    where TIn : ISample, new()
{
    if (input.Value % 3 == 0)
        return true;

    TIn newSample = new();
    newSample.Value = input.Value * 32;

    return Bar(newSample);
}

public bool Bar<TIn>(TIn input)
    where TIn : ISample =>
    input.Value % 5 == 0;
```

While redundant, the `new from` would certainly work with generic methods:

```cs
public bool Foo<TIn>(TIn input)
    where TIn : ISample
{
    if (input.Value % 3 == 0)
        return true;

    var collection = new from(List<>, typeof(TIn));
    for (int i = 0; i < input.Value; i++)
    {
        TIn newSample = new();
        newSample.Value = input.Value * i + 32;
        collection.Add(newSample);
    }

    return collection.All(Bar);
}
```

However, there are *tons* of questions and downsides to this approach that make me want to stay away from it. I'd much rather just leverage generics the way they were intended to be used.

**Comment by theunrepentantgeek** [comment]
The infrastructure to "recover" from having a non-generic value already exists with **dynamic** - here's an example that shows how you can go from something held in a variable of type `object` to a `List<T>` with the correct value for `T`:

``` csharp
    public void M() 
    {
        var item = 9;
        object obj = item;
        dynamic dyn = item;
        
        var listA = CreateList(item);
        Console.WriteLine(listA.GetType().ToString());
        
        var listB = CreateList(obj);
        Console.WriteLine(listB.GetType().ToString());
        
        var listC = CreateList(dyn);
        Console.WriteLine(listC.GetType().ToString());
    }
    
    private IEnumerable<T> CreateList<T>(T item) 
    {
        return new List<T>{ item };
    }
```

[SharpLab](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMAWAbjz1QGYT0iBhIgbzyM4cY6/dy6CiANwCGYIgGMiAXiJQApgHdmACgCUVAUM6SAdAFkNWwQF9enCyXqoyRI+qJX+OzmIkBLYAoC2sogCcJq4A9gBGAFYKksBE4RH+Xr7BOnAAnlCiPh7S6VCJ3j4pQlaC7kQANh4AzsAAgv5MYAqi3gAyNcCqST6apVykAapVtXV6AOIKwAAqaQAOChp60yEAysBgHlAA5hp92jr9buKVnQBCjc2tCh21qvH7riSYQyPAZxNTswtLK+ubOz2xUERxEJzeLDkTRa7U6qjyj1cg2GnSYnxm80W6mWaw2W126kRnHMBycpLmmzE3metAAPNMAHzMK6w2r0hmqaZEHqOZyg1AAdnkyiIt2A7NY3MKRFMxRJpiAA==)

HOWEVER, the problem is that you can't do very much with the list once you have it. Sure, it's a `List<T>`, but you can't treat it as a `List<T>`, you have to treat it as a non-generic type, say `IEnumerable`, which kinda defeats the point.

I've needed code just like that in the original post. And the fix was to propagate the generic types out from that point, so that you don't end up with `MyObject` of type `Object`, you end up with something of a definite type.

> Often you can't use generics because you don't know the type at compile time.

Generic library code that I write never knows the type at compile type. That's literally why I write generic code, so that it works even though I don't know the actual type that will be used.

**Comment by iam3yal** [comment]
I think that your example in the OP doesn't really represent the problem, in my experience the problem many times is with library authors that are exposing only the generic version of a type or a method and it may be totally appropriate but then consumers needs to deal with it in their edge cases and it's certainly shouldn't be a common practice **but** if it is, you either trying to use the wrong tool for the job or picked the wrong language for you to begin with so I really don't think the language needs to change because of that and I don't think something like this would ever make it into the language not only because it shouldn't be common but the reliance on reflection is detestable; however, you can always write utility or extension methods to ease your pain.

I'd love to have generic specialization, (a lot) more generic type inference, C++ concepts like feature for generics but dynamic? maybe in some other circumstances.

Referencing your old issue #2802 from 2019.
